Improve LSP0 and LSP14 and LSP17 Specification (#257)
lukso-network/LIPs: LSPs/LSP-0-ERC725Account.md

---
lip: 0
title: ERC725 Account
author: Fabian Vogelsteller <fabian@lukso.network>
discussions-to: https://discord.gg/E2rJPP4
status: Draft
type: LSP
created: 2021-09-21
requires: ERC165, ERC725X, ERC725Y, ERC1271, LSP1, LSP2, LSP14, LSP17, LSP20
---

## Simple Summary

The **ERC725Account** standard outlines a blockchain smart contract account that integrates a suite of standards, including [ERC725], to **represent and manage digital identities**.

## Abstract

> This standard defines a blockchain-based account system that can be used by humans, machines, organizations or other smart contracts.

Reviewer: ```suggestion
This standard defines a blockchain-based account system that humans, machines, organizations or other smart contracts can use.
```

Key functionalities of this account system include:

> - **Dynamic Information Attachment**: Leveraging [ERC725Y] for adding generic information to the account post-deployment.

Reviewer: I think using the gerund and suffixes like "ment" or "ing" makes the sentence and the specs quite hard to read. As the text is already quite formal.

Suggesting to re-word more with present tense:
(also there should not be capital letter after colons `:`)

For the last remaining points:
- for LSP17, not sure if we should mention interface IDs, but just the word **interface**. It's more generic and understandable for non-smart contract developers.
- for LSP20, I proposed re-wording as the sentence current sentence seems hard to read.

---

Key functionalities of this account system include:

- **Attach Information Dynamically**: leverages [ERC725Y] to add generic information to the account after-deployment.

- **Generic Execution**: utilises [ERC725X] to allow the account to interact with other contracts, manage token transfers, and initiate contract creation with different operations.

- **Signature Verification**: uses [ERC1271] to verify whether the signature on behalf of the account is valid.

- **Notifications and Reactivity**: with [LSP1-UniversalReceiver], enables the account to be aware of assets or any other information and react accordingly (e.g., denying specific tokens).

- **Secured Ownership Management**: enforced by [LSP14-Ownable2Step], ensures transfer of ownership is secured through a 2-step process.

- **Future-Proof Functionality**: through [LSP17-ContractExtension], the account can extend and support new standardised functions and interfaces over time.

- **Streamlined Account Interaction**: via [LSP20-CallVerification], to make interactions with the account easier. The contract can be called directly rather than through the contract that controls it (if the [`owner()`](#owner) is a contract).

- **Signature Verification**: Utilizing [ERC1271] to verify whether the signature on behalf of the account is valid.

- **Notifications and Reactivity**: Through [LSP1-UniversalReceiver], enabling the account to be aware of assets or any other information and react accordingly (e.g., denying specific tokens).

- **Secured Ownership Management**: Enforced by [LSP14-Ownable2Step], ensuring transfer of ownership is secured through a 2-step process.

- **Future-Proof Functionalitys**: Through [LSP17-ContractExtension], allowing the account to extend and support new standardized functions and interfaceIds over time.

- **Streamlined Account Interaction**: Via [LSP20-CallVerification], enabling direct function calls on the account from addresses other than the owner, with verification of the call occurring on the owner if it is a contract.

## Motivation

### Limitations of Externally Owned Accounts

Using Externally Owned Accounts (EOAs) as main accounts poses challenges in different aspects. The significant limitations include:

> - **No Information Attachment**: EOAs cannot store data that is readeable by off-chain clients or other smart contracts.

Reviewer: ```suggestion
- **No Information Attachment**: EOAs can not store data readable by off-chain clients or other smart contracts.
```

> - **Weak Security**:The account, represented by a public key is controlled by its corresponding private key; if the private key leaks or is lost, all associated assets, reputation, and control over other smart contracts are also lost.

Reviewer: ```suggestion
- **Weak Security**: the account, represented by a public key, is controlled by its corresponding private key. If the private key leaks or is lost, all associated assets, reputation, and control over other smart contracts are also lost.
```

> - **Exclusive Control**: Sharing the private key with other entities is not an option, as possession equals full control. This exclusivity prevents delegation of certain access rights, social recovery, and other shared security measures.

Reviewer: ```suggestion
- **Exclusive Control**: sharing the private key with other entities is not an option, as possession equals full control. This exclusivity prevents delegating certain access rights, social recovery, and other shared security measures.
```

> - **Information Tracking Limitation**: EOAs do not allow for the internal tracking of asset transfers, followers, or other information. As a result, account holders must depend on external blockchain explorers to monitor their transaction history, which introduce inconvenience and slow down information access.

Reviewer: ```suggestion
- **Information Tracking Limitation**: EOAs do not allow for the internal tracking of asset transfers, followers, or other information. As a result, account holders must depend on external blockchain explorers to monitor their transaction history, which introduces inconvenience and slows down information access.
```

### Lack of Standardization of Accounts

The absence of a standardized specification for blockchain accounts presents several challenges:

- **Complex Development Landscape**: The lack of a standard account template in blockchain leads to a diverse and complex array of custom smart contract accounts created by individual developers and organizations, which make it harder for other developers to build protocols on top of it.

- **Adoption and Compatibility Hurdles**: This diversity necessitates extra compatibility work, slowing the adoption and practical use of smart contract accounts across the blockchain ecosystem.

> ### Advantages of Smart Contract Accounts

Reviewer: ```suggestion
### Advantages of Smart Contract Based Accounts
```

Adopting smart contract accounts over traditional EOAs streamlines blockchain infrastructure use. Smart contracts enable advanced features and enhanced control, such as:

> - **Dynamic Information Storage**: Unlike EOAs, smart contract accounts can record static information, making them self-explanatory entities that can assist in identity verification for off-chain entities, and enable the decentralized tracking of any information stored such as assets, followers, etc. The information stored can also be standardized, for example [LSP3-Profile-Metadata] for storing profile information, [LSP5-ReceivedAssets] for storing the addresses of received assets, and [LSP10-ReceivedVaults] for storing addresses of received vaults, etc ..

Reviewer: ```suggestion
- **Dynamic Information Storage**: unlike EOAs, smart contract accounts can record static information, making them self-explanatory entities that can assist in identity verification for off-chain entities and enable the decentralized tracking of any information stored, such as assets, followers, etc. The information stored can also be standardized, for example, [LSP3-Profile-Metadata] for storing profile information, [LSP5-ReceivedAssets] for storing the addresses of received assets, and [LSP10-ReceivedVaults] for storing addresses of received vaults, etc...
```

> - **Upgradeable Security**: Owners can use a simple EOA for owning the account having basic level of security, or opt for an owner having a multisignature setup, or permissions access control contracts providing advanced layer of security in contorlling the account.

Reviewer: ```suggestion
- **Upgradeable Security**: owners can use a simple EOA for owning the account having a basic level of security or opt for an owner having a multi-signature setup or permissions access control contracts providing an advanced layer of security in controlling the account.
```

> - **Inclusive Control**: Users can decide to own the account with a permission based access control contract (e.g: the [LSP6-KeyManager]) which allows owners to assign specific rights to different actors, like granting a party the right to update profile information without providing full transactional authority.

Reviewer: ```suggestion
- **Inclusive Control**: users can decide to own the account with a permission-based access control contract (e.g., the [LSP6-KeyManager]), which allows owners to assign specific rights to different actors, like granting a party the right to update profile information without providing full transactional authority.
```
